### ISSUE TITLE
Problem with STB support

### ISSUE THREAD

**Issue by jaja360** [mlpack/mlpack#3514]
#### Issue description

Since on my system (Ubuntu 23.04) the `mlpack-dev` package does not seem available, I installed the library using cmake by following the steps in [the readme file](https://github.com/mlpack/mlpack#3-installing-and-using-mlpack-in-c). Everything installs with no error. However, when I try to load an image with the `data::Load` function, I always get "false" and the image is not loaded.

#### Your environment

 * version of mlpack: latest at the time of writing this (5e8beeb0b)
 * operating system: Ubuntu 23.04
 * compiler: g++ 12.2
 * version of dependencies (Boost/Armadillo): (Armadillo: 11.4.2, Ensmallen: 2.19.1, Cereal: 1.3.2)
 * any other environment information you think is relevant:

When doing `cmake ..`, the output contains the following lines:
```
-- Found StbImage: /usr/include/stb/
-- Check that STB static implementation mode links correctly...
-- Check that STB static implementation mode links correctly... success
```
According to the readme, "If the STB library headers are available, image loading support will be available.", therefore I assume I don't need to specify a flag manually when compiling the library with `cmake ..` and `sudo make install`. 

#### Steps to reproduce

I compile this simple program:
```c++
#include <mlpack.hpp>
using namespace mlpack;
int main() {
  data::ImageInfo info;
  arma::mat matrix;
  data::Load("test.bmp", matrix, info, true);
}
```
using this command: `g++ -std=c++17 -I/usr/include/stb/ -I/usr/local/include/ test.cpp -larmadillo -L/usr/lib/gcc/x86_64-linux-gnu/12/ -lgomp -lpthread` (I used the flags given by `pkg-config --cflags mlpack` and `pkg-config --libs mlpack`).

#### Expected behavior

I expect no runtime error.

#### Actual behavior

I receive the following error:
```
[FATAL] Load(): mlpack was not compiled with STB support, so images cannot be loaded!

terminate called after throwing an instance of 'std::runtime_error'
  what():  fatal error; see Log::Fatal output
```

**Comment by jaja360** [comment]
Update: I also tried to install mlpack on a different computer, running Ubuntu 22.04 LTS. On this computer, the package libmlpack-dev is available but is a really old version (3.4.2), so I tried to compile mlpack instead. I get the same error message as on my other computer, "mlpack was not compiled with STB support", even though cmake finds STB correctly.

**Comment by kiner-shah** [comment]
- Since you are building from source code, you can do some debugging to find any potential problems.
- The log message that you get is from "src/mlpack/core/data/load_image_impl.hpp" line 177 (at least in mlpack 4.2.0)
- The message gets printed when the pre-processor flag `HAS_STB` isn't set.
- This pre-processor flag is set in "CMakeLists.txt" line 322. You can add a line after line 322: `message(STATUS "STB_IMAGE_FOUND is set and pre-processor flag HAS_STB has been added")`.
- Note that `STB_IMAGE_FOUND` is set during `find_package(StbImage)` which should be above that line in the same file.
- `find_package` will call "CMake/FindStbImage.cmake". You can put similar debug messages in this file to find any problems.

**Comment by jaja360** [comment]
Thanks for you answer,

When I add `message(STATUS "STB_IMAGE_FOUND is set and pre-processor flag HAS_STB has been added")` after line 322, I correctly see it displayed when running `cmake ..`. The mlpack CLI apps seems to be compiled correctly with the `-DHAS_STB` flag.

I'm not much familiar with cmake, so maybe I just don't understand what it is supposed to do, but I would expect the `#define HAS_STB` to be included by cmake in the mlpack headers installed in my `/usr/local/include/mlpack` folder, so that when I include mlpack in my c++ programs, the version that supports STB is loaded. For now, it doesn't seem the case. I managed however to make it work by manually adding the `#define HAS_STB` before the `#include <mlpack.hpp>` in the test program above.

Is that what we are supposed to do (defining HAS_STB manually in our programs before including mlpack) ?
If yes, I think the install instructions and the [Image Utilities Tutorial](https://github.com/mlpack/mlpack/blob/4.2.0/doc/tutorials/image.md) should mention this. I guess another solution would be for cmake to add `-DHAS_STB` in the `cflags` of the pkg-config file it generates.

**Comment by rcurtin** [comment]
Thanks for the report.  This is actually a leftover from before mlpack changed to be a header-only library.  You are right that the workaround for now is just to add `-DHAS_STB` to your compilation command, but you are also right that this should be automatically defined at install time whenever STB is detected.

It may be a couple days before I get to it, but I'll open a PR that handles this, likely by generating a file with `HAS_STB` set or unset based on what is found at configuration/build time.

(I'll also probably rename it `MLPACK_USE_STB` or similar, so it doesn't interfere with any other symbols.  Originally, when mlpack was not header-only, the macro was only used inside of libmlpack.so and thus never needed to be encountered by the user.  However, this was not properly adapted when mlpack was changed to be header-only, like I alluded to earlier.)

**Comment by jaja360** [comment]
Sounds good, thanks !

**Comment by rcurtin** [comment]
I opened a fix in #3529.  I think the description of the fix may be longer than the fix itself, but I'm not sure.  The tl;dr is that mlpack's next release will have a file `mlpack/config.hpp` that automatically defines `MLPACK_HAS_STB` depending on whether STB was found when mlpack was configured and installed.  (Also, the macro was renamed to avoid conflicts.)

**Comment by rcurtin** [comment]
Okay, with #3529 merged, I think that this issue is fixed.  Feel free to try on the current master branch and report back if there are still issues, or wait until the next stable release (should be soon).  Thanks again for the report!
